ssc/Pos/Ssc/GState.hs

```diff
@@ -0,0 +1,342 @@
+                                                   verifyAndApplySscPayload)
+import qualified Pos.Ssc.GodTossing.VssCertData   as VCD
+import           Pos.Ssc.Lrc                      (getSscRichmenFromLrc)
+import           Pos.Ssc.Mem                      (MonadSscMem, SscGlobalQuery,
+                                                   SscGlobalUpdate, askSscMem,
+                                                   sscRunGlobalQuery, sscRunGlobalUpdate)
+import           Pos.Ssc.SeedError                (SscSeedError)
+import           Pos.Ssc.Types                    (SscBlock (..), SscGlobalState (..),
+                                                   sgsCommitments, sgsOpenings, sgsShares,
+                                                   sgsVssCertificates, sscGlobal)
+import           Pos.Ssc.VerifyError              (SscVerifyError (..),
+                                                   sscIsCriticalError)
+import           Pos.Util.Chrono                  (NE, NewestFirst (..), OldestFirst (..))
+import           Pos.Util.Util                    (inAssertMode, _neHead, _neLast)
+
+----------------------------------------------------------------------------
+-- Utilities
+----------------------------------------------------------------------------
+
+gtGetGlobalState
+    :: (MonadSscMem ctx m, MonadIO m)
+    => m SscGlobalState
+gtGetGlobalState = sscRunGlobalQuery ask
+
+getGlobalCerts
+    :: (MonadSscMem ctx m, MonadIO m)
+    => SlotId -> m VssCertificatesMap
+getGlobalCerts sl =
+    sscRunGlobalQuery $
+        VCD.certs .
+        VCD.setLastKnownSlot sl <$>
+        view sgsVssCertificates
+
+-- | Get stable VSS certificates for given epoch.
+getStableCerts
+    :: (HasGtConfiguration, HasConfiguration, MonadSscMem ctx m, MonadIO m)
+    => EpochIndex -> m VssCertificatesMap
+getStableCerts epoch =
+    getStableCertsPure epoch <$> sscRunGlobalQuery (view sgsVssCertificates)
+
+----------------------------------------------------------------------------
+-- Queries & updates
+----------------------------------------------------------------------------
+
+-- | Calculate 'SharedSeed' for given epoch using 'SscGlobalState'.
+sscCalculateSeed
+    :: forall ctx m.
+       ( MonadSscMem ctx m
+       , MonadDBRead m
+       , MonadReader ctx m
+       , HasLrcContext ctx
+       , MonadIO m
+       , WithLogger m )
+    => EpochIndex
+    -> m (Either SscSeedError SharedSeed)
+sscCalculateSeed epoch = do
+    -- We take richmen for the previous epoch because during N-th epoch we
+    -- were using richmen for N-th epoch for everything – so, when we are
+    -- calculating the seed for N+1-th epoch, we should still use data from
+    -- N-th epoch.
+    richmen <- getSscRichmenFromLrc "sscCalculateSeed" (epoch - 1)
+    sscRunGlobalQuery $ sscCalculateSeedQ epoch richmen
+
+sscCalculateSeedQ
+    :: EpochIndex
+    -> RichmenStakes
+    -> SscGlobalQuery (Either SscSeedError SharedSeed)
+sscCalculateSeedQ _epoch richmen =
```

Comment:
I would delete `_epoch` argument, because it was here only because of nist beacon.